python/nx-cugraph/nx_cugraph/algorithms/traversal/breadth_first_search.py

```diff
@@ -0,0 +1,256 @@
+    distances, predecessors, node_ids = plc.bfs(
+        handle=plc.ResourceHandle(),
+        graph=G._get_plc_graph(switch_indices=reverse),
+        sources=cp.array([src_index], dtype=index_dtype),
+        direction_optimizing=False,
+        depth_limit=-1 if depth_limit is None else depth_limit,
+        compute_predecessors=True,
+        do_expensive_check=False,
+    )
+    mask = predecessors >= 0
+    return distances[mask], predecessors[mask], node_ids[mask]
+
+
+@networkx_algorithm
+def generic_bfs_edges(G, source, neighbors=None, depth_limit=None, sort_neighbors=None):
+    """`neighbors` parameter is not yet supported."""
+    return bfs_edges(source, depth_limit=depth_limit)
+
+
+@generic_bfs_edges._can_run
+def _(G, source, neighbors=None, depth_limit=None, sort_neighbors=None):
+    return neighbors is None and sort_neighbors is None
+
+
+@networkx_algorithm
+def bfs_edges(G, source, reverse=False, depth_limit=None, sort_neighbors=None):
+    """`sort_neighbors` parameter is not yet supported."""
+    G = _check_G_and_source(G, source)
+    if depth_limit is not None and depth_limit < 1:
+        return
+    distances, predecessors, node_ids = _bfs(
+        G, source, depth_limit=depth_limit, reverse=reverse
+    )
+    # Using groupby like this is similar to bfs_predecessors
+    groups = _groupby([distances, predecessors], node_ids)
+    id_to_key = G.id_to_key
+    for key in sorted(groups):
+        children_ids = groups[key]
+        parent_id = key[1]
+        parent = id_to_key[parent_id] if id_to_key is not None else parent_id
+        yield from zip(
+            repeat(parent, children_ids.size),
+            G._nodeiter_to_iter(children_ids.tolist()),
+        )
+
+
+@bfs_edges._can_run
+def _(G, source, reverse=False, depth_limit=None, sort_neighbors=None):
+    return sort_neighbors is None
+
+
+@networkx_algorithm
+def bfs_tree(G, source, reverse=False, depth_limit=None, sort_neighbors=None):
+    """`sort_neighbors` parameter is not yet supported."""
+    G = _check_G_and_source(G, source)
+    if depth_limit is not None and depth_limit < 1:
+        return nxcg.DiGraph.from_coo(
+            1,
+            cp.array([], dtype=index_dtype),
+            cp.array([], dtype=index_dtype),
+            id_to_key=[source],
+        )
+
+    distances, predecessors, node_ids = _bfs(
+        G,
+        source,
+        depth_limit=depth_limit,
+        reverse=reverse,
+    )
+    if predecessors.size == 0:
+        return nxcg.DiGraph.from_coo(
+            1,
+            cp.array([], dtype=index_dtype),
+            cp.array([], dtype=index_dtype),
+            id_to_key=[source],
+        )
+    # TODO: create renumbering helper function(s)
+    unique_node_ids = cp.unique(cp.hstack((predecessors, node_ids)))
+    # Renumber edges
+    # Option 1
+    src_indices = cp.searchsorted(unique_node_ids, predecessors)
+    dst_indices = cp.searchsorted(unique_node_ids, node_ids)
+    # Option 2
```

Comment:
Why are we keeping the commented-out option 2 code?

Reply:
Chose option 1 and removed option 2. Perhaps we can revisit option 2 once we make renumbering helper functions.